File: priv/html_sanitizer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys
import struct
import re
from BeautifulSoup import BeautifulSoup, NavigableString
import argparse


parser = argparse.ArgumentParser(description='Sanitize HTML.')
parser.add_argument('--oformat', nargs='?', default="html",
                    dest="oformat",
                    help='output format: [html | text]')

args = parser.parse_args()

acceptable_elements = ['a', 'abbr', 'acronym', 'address', 'area', 'b', 'big',
                       'blockquote', 'br', 'button', 'caption', 'center', 'cite', 'code', 'col',
                       'colgroup', 'dd', 'del', 'dfn', 'dir', 'div', 'dl', 'dt', 'em',
                       'font', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hr', 'i', 'img',
                       'ins', 'kbd', 'label', 'legend', 'li', 'map', 'menu', 'ol',
                       'p', 'pre', 'q', 's', 'samp', 'small', 'span', 'strike',
                       'strong', 'sub', 'sup', 'table', 'tbody', 'td', 'tfoot', 'th',
                       'thead', 'tr', 'tt', 'u', 'ul', 'var']

acceptable_attributes = ['abbr', 'accept', 'accept-charset', 'accesskey',
                         'action', 'align', 'alt', 'axis', 'border', 'cellpadding', 'cellspacing',
                         'char', 'charoff', 'charset', 'checked', 'cite', 'clear', 'cols',
                         'colspan', 'color', 'compact', 'coords', 'datetime', 'dir',
                         'enctype', 'for', 'headers', 'height', 'href', 'hreflang', 'hspace',
                         'id', 'ismap', 'label', 'lang', 'longdesc', 'maxlength', 'method',
                         'multiple', 'name', 'nohref', 'noshade', 'nowrap', 'prompt',
                         'rel', 'rev', 'rows', 'rowspan', 'rules', 'scope', 'shape', 'size',
                         'span', 'src', 'start', 'style', 'summary', 'tabindex', 'target', 'title', 'type',
                         'usemap', 'valign', 'value', 'vspace', 'width']

if args.oformat == "text":
    acceptable_elements = []
    acceptable_attributes = []
elif args.oformat == "html":
    pass
elif args.oformat == "markdown":
    import markdown

BeautifulSoup.QUOTE_TAGS = ['pre', 'code']

def clean_html( fragment, quirks=True ):
    while True:
        soup = BeautifulSoup( fragment )
        removed = False
        for tag in soup.findAll(True): # find all tags
            if tag.name == 'script':
                if re.match("https?:\/\/(?:www\.)?gist\.github\.com\/\d+\.js", dict(tag.attrs)[u'src']):
                    removed = False
                if re.match("(https?:)?\/\/speakerdeck.com\/assets\/embed\.js", dict(tag.attrs)[u'src']):
                    removed = False
                else:
                    tag.extract()
                    removed = True
            elif tag.name == 'iframe':
                if re.match("https?:\/\/(?:www\.)?youtube(?:-nocookie)?\.com\/", dict(tag.attrs)[u'src']):
                    removed = False
                elif re.match("https?:\/\/(?:www\.)?slideshare\.net\/slideshow\/embed_code\/\d+", dict(tag.attrs)[u'src']):
                    removed = False
                else:
                    tag.extract()
                    removed = True
            elif tag.name == 'mt-cut':
                removed = False
            elif tag.name == 'pre':
                pass
            elif tag.name == 'div' and dict(tag.attrs)[u'class'] == 'codehilite':
                removed = False
            elif tag.name not in acceptable_elements:
                tag.extract() # remove the bad ones
                removed = True
            else: # it might have bad attributes
                if not len(tag.contents):
                    pass
                elif quirks:
                    stopNode = tag.contents[-1]
                    strings = []
                    current = tag.contents[0]
                    while current:
                        if not isinstance(current, NavigableString):
                            text = str(current)
                            pass
                        else:
                            text = current
                            text = re.sub(re.compile('\n[\n]+', re.UNICODE), '</p><p>', text)
                            text = re.sub(re.compile('\n', re.UNICODE), '<br />', text)
                            current.replaceWith(text)
                        current = current.next

                # a better way to get all attributes?
                for attr in tag._getAttrMap().keys():
                    if attr not in acceptable_attributes:
                        del tag[attr]

        # turn it back to html
        fragment = unicode(soup)

        if removed:
            # we removed tags and tricky can could exploit that!
            # we need to reparse the html until it stops changing
            continue # next round

        return fragment

def clean_text(text):
    text = re.sub(re.compile('</?p>', re.UNICODE), '\n', text)
    text = re.sub(re.compile('<br(\s*/?\s*)?>', re.UNICODE), '\n', text)
    text = re.sub(re.compile('<.*?>', re.UNICODE), '', text)
    return text

inlen = struct.unpack('>I', sys.stdin.read(4))
intext = unicode(sys.stdin.read(inlen[0]), "utf-8")

if args.oformat == "html":
    outtext = clean_html(intext).encode('utf-8')
elif args.oformat == "markdown":
    outtext = clean_html(markdown.markdown(intext, ['codehilite']), quirks=False).encode('utf-8')
elif args.oformat == "text":
    outtext = clean_text(intext).encode('utf-8')

outlen = len(outtext)

sys.stdout.write(struct.pack('>I', outlen))
sys.stdout.write(outtext)

File: priv/html_sanitizer.rb
#!/usr/bin/ruby
# Encoding: utf-8
require 'rubygems'
require 'sanitize'

def errlog(str)
  File.open('/tmp/san.log', 'a') do |f|
    f.puts str
  end
end

class HTMLSanitize
  module Config
    RELAXED = {
      :elements => %w[
        a abbr b bdo blockquote br caption cite code col colgroup dd del dfn dl
        dt em figcaption figure h1 h2 h3 h4 h5 h6 hgroup i img ins kbd li mark
        ol p pre q rp rt ruby s samp small strike strong sub sup table tbody td
        tfoot th thead time tr u ul var wbr
      ],

      :attributes => {
        :all         => ['dir', 'lang', 'title'],
        'a'          => ['href'],
        'blockquote' => ['cite'],
        'col'        => ['span', 'width'],
        'colgroup'   => ['span', 'width'],
        'del'        => ['cite', 'datetime'],
        'img'        => ['align', 'alt', 'height', 'src', 'width'],
        'ins'        => ['cite', 'datetime'],
        'ol'         => ['start', 'reversed', 'type'],
        'q'          => ['cite'],
        'table'      => ['summary', 'width'],
        'td'         => ['abbr', 'axis', 'colspan', 'rowspan', 'width'],
        'th'         => ['abbr', 'axis', 'colspan', 'rowspan', 'scope', 'width'],
        'time'       => ['datetime', 'pubdate'],
        'ul'         => ['type']
      },

      :protocols => {
        'a'          => {'href' => ['ftp', 'http', 'https', 'mailto', :relative]},
        'blockquote' => {'cite' => ['http', 'https', :relative]},
        'del'        => {'cite' => ['http', 'https', :relative]},
        'img'        => {'src'  => ['http', 'https', :relative]},
        'ins'        => {'cite' => ['http', 'https', :relative]},
        'q'          => {'cite' => ['http', 'https', :relative]}
      },
      :transformers =>
      [
       lambda do |env|
         node      = env[:node]
         node_name = env[:node_name]
         return if env[:is_whitelisted] || !node.element?
         return unless node_name == 'iframe'
         return unless node['src'] =~ /\Ahttps?:\/\/(?:www\.)?youtube(?:-nocookie)?\.com\//
         Sanitize.clean_node!(node, {
                                :elements => %w[iframe],
                                :attributes => {
                                  'iframe'  => %w[allowfullscreen frameborder height src width]
                                }
                              })
         {:node_whitelist => [node]}
       end,

       lambda do |env|
         node      = env[:node]
         node_name = env[:node_name]
         return if env[:is_whitelisted] || !node.element?
         return unless node_name == 'script'
         return unless node['src'] =~ /\Ahttps?:\/\/(?:www\.)?gist\.github\.com\/\d+\.js/
         Sanitize.clean_node!(node, {
                                :elements => %w[script],
                                :attributes => {
                                  'script'  => %w[src]
                                }
                              })
         {:node_whitelist => [node]}
       end
      ]
    }
  end
end

errlog("Start")

inlen = STDIN.read(4).unpack('N')
intext = STDIN.read(inlen[0])

outtext = Sanitize.clean(intext, HTMLSanitize::Config::RELAXED)
outlen = outtext.size()

STDOUT.write([outlen].pack('N'))
STDOUT.write(outtext)
